### Importing Pandas

#### Packages
Packages provide additional tools and functions not present in base Python. Python includes a number of packages to start with, the Anaconda distribution which we've all downloaded for Unit 3 comes with the "Pandas" package already installed.

Once you've installed a package, you can load it into your current Python session with the import function. Otherwise these functions will not be available.


#### Pandas

Like spreadsheets in Microsoft Excel, Pandas allows us to store our data in tabular, multi-dimensional objects (dataframes) with familiar features like rows, columns, and headers. This is useful because it makes management, manipulation, and cleaning of large datasets much easier than would be the case using Python's built-in data structures such as lists. Pandas also provides a wide range of useful tools for working with data once it has been stored and structured.

Begin by importing the pandas package using the following command:

In [2]:
import numpy as np
import pandas as pd

### Creating a DataFrame

#### Working Directories & Relative Paths

By now, you should have either downloaded the csv file "Apple.csv" from canvas, or saved your own data as a csv file. I've stored my copy in the same folder as this Jupyter Notebook. **NOTE:** make sure that your csv file is saved in the same working directory as your .ipynb notebook file that you will use. 

Remember that Jupyter Notebooks automatically set your working directory to the folder where the .ipynb is saved. You'll have to save the document at least once to set your directory, but once there you can use what's called relative file paths to access the files there.

If a file is located in your working directory, its relative path is just the name of the file!

#### Using the `pd.read_csv()` function

`pd.read_csv` reads the tabular data from a Comma Separated Values (csv) file into a dataframe object that we'll define as `df`.

To create our dataframe object we'll define our object `df` by executing the `pd.read_csv()`function on our data file by inserting the relative file path into the parathenses.

In [3]:
df=pd.read_csv("Apple.csv")

### Finding and Filtering Data
You are going to explore the Apple Stock data set and go through logical steps to arrive at the number of years that the Apple share price was over 70$. But firstly, you should try and get an idea of what the data is about so you can understand the content and structure of this data set. The first step you are going to do is take samples from the chosen data set.You will type "df.sample(n=x)" and that would give you an x amount of random samples of the data frame. The x represents the number of samples and in this example, try  the number 40 for x.

In [107]:
df.sample(n=40)

Date       Open       High        Low      Close  Adj Close  \
8140  2013-03-22  16.235001  16.503571  16.182501  16.496786  14.465927   
4016  1996-10-30   0.209821   0.214286   0.204241   0.204241   0.176493   
1127  1985-05-29   0.076451   0.077009   0.076451   0.076451   0.060316   
6817  2007-12-18   6.661428   6.690357   6.378572   6.535000   5.647151   
566   1983-03-10   0.194754   0.196987   0.190290   0.191964   0.151451   
1178  1985-08-09   0.068080   0.068080   0.068080   0.068080   0.053712   
6031  2004-11-03   0.970893   1.001964   0.964107   0.987679   0.853492   
8575  2014-12-11  28.065001  28.450001  27.834999  27.905001  25.463926   
3618  1995-04-05   0.304688   0.310268   0.301339   0.310268   0.265819   
2900  1992-06-03   0.504464   0.504464   0.482143   0.483259   0.400285   
3294  1993-12-22   0.243304   0.254464   0.241071   0.250000   0.210634   
3664  1995-06-12   0.392857   0.397321   0.391741   0.394392   0.338828   
4825  2000-01-14   0.892857   0.912946   0.887277   0.896763   0.774928   
4467  1998-08-14   0.363281   0.363839   0.352679   0.361607   0.312479   
4414  1998-06-01   0.236607   0.246652   0.228795   0.234375   0.202533   
2632  1991-05-13   0.466518   0.477679   0.459821   0.470982   0.385817   
4893  2000-04-24   1.026786   1.075893   1.024554   1.075893   0.929721   
954   1984-09-20   0.121094   0.122210   0.121094   0.121094   0.095538   
9940  2020-05-15  75.087502  76.974998  75.052498  76.927498  76.789047   
7084  2009-01-09   3.328928   3.335000   3.219286   3.235000   2.795490   
6677  2007-05-31   4.288214   4.363214   4.269286   4.328214   3.740180   
5915  2004-05-19   0.489286   0.491071   0.471786   0.472679   0.408460   
1067  1985-03-04   0.112723   0.116071   0.112723   0.112723   0.088934   
6755  2007-09-20   5.005357   5.063929   4.975714   5.011071   4.330263   
5084  2001-01-25   0.367188   0.367188   0.352679   0.356027   0.307657   
1414  1986-07-18   0.143973   0.145089   0.139509   0.141741   0.111827   
3206  1993-08-18   0.258929   0.265625   0.252232   0.254464   0.213629   
7803  2011-11-15  13.600000  13.910714  13.551785  13.886786  12.000114   
5645  2003-04-24   0.241429   0.243036   0.232143   0.240000   0.207393   
6798  2007-11-20   5.916786   6.135357   5.840357   6.030357   5.211069   
4947  2000-07-11   1.017857   1.058036   0.989955   1.016741   0.878606   
7649  2011-04-07  12.075000  12.158215  12.001071  12.074286  10.433864   
6931  2008-06-03   6.673572   6.721428   6.512143   6.620357   5.720910   
2812  1992-01-28   0.578125   0.583705   0.562500   0.582589   0.480694   
9963  2020-06-18  87.852501  88.362503  87.305000  87.932503  87.774246   
7968  2012-07-13  21.533928  21.685358  21.428572  21.606071  18.670649   
1159  1985-07-15   0.079799   0.081473   0.079241   0.079241   0.062518   
1610  1987-04-28   0.338170   0.347656   0.337054   0.343750   0.271204   
3607  1995-03-21   0.316964   0.328125   0.314732   0.323661   0.277294   
1041  1985-01-24   0.132254   0.132254   0.129464   0.129464   0.102142   

            Volume  
8140  3.951052e+08  
4016  2.570512e+08  
1127  2.465568e+08  
6817  1.222603e+09  
566   1.126048e+08  
1178  6.095040e+07  
6031  1.204174e+09  
8575  1.656068e+08  
3618  2.648576e+08  
2900  3.005744e+08  
3294  1.813728e+08  
3664  2.121168e+08  
4825  3.903760e+08  
4467  4.507776e+08  
4414  3.196928e+08  
2632  2.449440e+08  
4893  4.436208e+08  
954   6.616960e+07  
9940  1.663484e+08  
7084  5.468456e+08  
6677  1.297066e+09  
5915  3.755920e+08  
1067  1.531040e+08  
6755  6.918408e+08  
5084  4.897088e+08  
1414  3.091200e+08  
3206  1.887200e+08  
7803  4.308108e+08  
5645  3.251080e+08  
6798  1.543643e+09  
4947  3.578960e+08  
7649  3.734472e+08  
6931  7.505204e+08  
2812  1.737232e+08  
9963  9.682040e+07  
7968  3.114272e+08  
1159  7.768320e+07  
1610  3.248000e+08  
3607  3.053680e+08  
1041  3.970624e+08

Now you started seeing some numbers such as Apple's open stock price and the volume. Something which is very important for stock traders is the price it started trading at and the price it is at now. You will start by filtering out the first 10 columns of the stock price and you will type "print(df.head(3))" . The word "head" in your code basically means the beginning of your data frame. And you probably already guessed it, if "head" shows you the beginning then "tail" shows you the end. We will start by "print(df.head(3))" then try "print(df.tail(3))" .

In [119]:
print(df.head(3))

         Date      Open      High       Low     Close  Adj Close       Volume
0  1980-12-12  0.128348  0.128906  0.128348  0.128348   0.101261  469033600.0
1  1980-12-15  0.122210  0.122210  0.121652  0.121652   0.095978  175884800.0
2  1980-12-16  0.113281  0.113281  0.112723  0.112723   0.088934  105728000.0


In [109]:
print(df.tail(3))

             Date        Open        High         Low       Close   Adj Close  \
10013  2020-08-28  126.012497  126.442497  124.577499  124.807503  124.807503   
10014  2020-08-31  127.580002  131.000000  126.000000  129.039993  129.039993   
10015  2020-09-01  132.759995  134.800003  130.529999  134.179993  134.179993   

            Volume  
10013  187630000.0  
10014  225702700.0  
10015  151948100.0  


We can notice that over 40 years the stock increased from a high of 0.1289 in 1980 and reached a high of 134.80 in 2020. That is an increase of 1,044%. 

Apple stock is one of the most important stocks in the US market as Apple is the biggest company is the world and there is a huge correlation between the market and the stock. So now that we found out Apple is at 134$ in 2020, how long was Apple able to sustaina price over 12 Was it a one day rise? One month? Or was it a year?

You will figure that out by inputting "print(df["High"] >= 70)". This code is telling pyhton to look through the columns under "High" and see if the price was greater than or equal to 70.

In [115]:
print(df["High"] >= 110)

0        False
1        False
2        False
3        False
4        False
         ...  
10011     True
10012     True
10013     True
10014     True
10015     True
Name: High, Length: 10016, dtype: bool



### Weird! Notice anything strange?
### Why did you get it in this format and not just get rows where the "High is greater than 70?
This is because this code is running a function that compares every single value in "High" to the number of "70" or greater and then assigning a true or false. To get what you want, you will have to to wrap the whole function "(df["High"] >= 70)" and add a df[ before it. When doing that you are referencing the High to your data frame and you end up with function print(df[(df["High"] >= 120)])
    

In [4]:
print(df[(df["High"] >= 70)])

             Date        Open        High         Low       Close   Adj Close  \
9836   2019-12-16   69.250000   70.197502   69.245003   69.964996   69.485619   
9837   2019-12-17   69.892502   70.442497   69.699997   70.102501   69.622192   
9838   2019-12-18   69.949997   70.474998   69.779999   69.934998   69.455833   
9839   2019-12-19   69.875000   70.294998   69.737503   70.004997   69.525352   
9840   2019-12-20   70.557503   70.662498   69.639999   69.860001   69.381348   
...           ...         ...         ...         ...         ...         ...   
10011  2020-08-26  126.180000  126.992500  125.082497  126.522499  126.522499   
10012  2020-08-27  127.142502  127.485001  123.832497  125.010002  125.010002   
10013  2020-08-28  126.012497  126.442497  124.577499  124.807503  124.807503   
10014  2020-08-31  127.580002  131.000000  126.000000  129.039993  129.039993   
10015  2020-09-01  132.759995  134.800003  130.529999  134.179993  134.179993   

            Volume  
9836  

In [7]:
Price_Filter=df[(df["High"] >= 70)]

### Results!
We are now finally able to see that Apple stock has been equal or greater than 70 for only one year which is in 2020. 

This indicates that Apple even though consider a safe stock is very volatile.

Since Apple and the stock market are strongly correlated, you can assume that the stock market behaves in a similar matter!

But why assume? Since you went through this model, go and find a date base of the US Stock market and filter the data just like we did now! 

In [12]:
Price_Filter.to_csv("Price_Filter.csv", index=False)